lib/spack/spack/fetch_strategy.py

```diff
@@ -262,13 +262,273 @@ def __init__(self, url=None, checksum=None, **kwargs):
 
         self.expand_archive = kwargs.get('expand', True)
         self.extra_options = kwargs.get('fetch_options', {})
-        self._curl = None
 
         self.extension = kwargs.get('extension', None)
 
         if not self.url:
             raise ValueError("URLFetchStrategy requires a url for fetching.")
 
+    def source_id(self):
+        return self.digest
+
+    def mirror_id(self):
+        if not self.digest:
+            return None
+        # The filename is the digest. A directory is also created based on
+        # truncating the digest to avoid creating a directory with too many
+        # entries
+        return os.path.sep.join(
+            ['archive', self.digest[:2], self.digest])
+
+    @property
+    def candidate_urls(self):
+        return [self.url] + (self.mirrors or [])
+
+    @_needs_stage
+    def fetch(self):
+        if self.archive_file:
+            tty.debug('Already downloaded {0}'.format(self.archive_file))
+            return
+
+        url = None
+        errors = []
+        for url in self.candidate_urls:
+            if not self._existing_url(url):
+                continue
+
+            try:
+                partial_file, save_file = self._fetch_from_url(url)
+                if save_file:
+                    os.rename(partial_file, save_file)
+                break
+            except FetchError as e:
+                errors.append(str(e))
+
+        for msg in errors:
+            tty.debug(msg)
+
+        if not self.archive_file:
+            raise FailedDownloadError(url)
+
+    def _existing_url(self, url):
+        tty.debug('Checking existence of {0}'.format(url))
+        # Telling urllib to check if url is accessible
+        try:
+            url, headers, response = web_util.read_from_url(url)
+        except web_util.SpackWebError:
+            msg = "Urllib fetch failed to verify url {0}".format(url)
+            raise FailedDownloadError(url, msg)
+        return (response.getcode() is None or response.getcode() == 200)
+
+    def _fetch_from_url(self, url):
+        save_file = None
+        partial_file = None
+        if self.stage.save_filename:
+            save_file = self.stage.save_filename
+            partial_file = self.stage.save_filename + '.part'
+        tty.msg('Fetching {0}'.format(url))
+
+        connect_timeout = spack.config.get('config:connect_timeout', 10)
+
+        timeout = self.extra_options.get('timeout')
+        if timeout:
+            connect_timeout = max(connect_timeout, int(timeout))
+
+        # Run urllib but grab the mime type from the http headers
+        try:
+            url, headers, response = web_util.read_from_url(url)
+        except web_util.SpackWebError as e:
```

Comment:
I haven't done a full review, but you can bump up the test coverage by covering this code-block. Try running with a context manager that makes all url requests fail.

@loulawrence